File: python/leetcode/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            if i == 0 or nums[i-1] != nums[i]:
                self.twoSum(nums, i, res)
        return res

    def twoSum(self, nums, i, result):
        left, right = i+1, len(nums)-1

        while left < right:
            if nums[left] + nums[right] + nums[i] == 0:
                result.append([nums[left], nums[right], nums[i]])
                left += 1
                right -= 1
                while left < right and nums[left] == nums[left-1]:
                    left += 1
            elif nums[left] + nums[right] + nums[i] > 0:
                right -= 1
            else:
                left += 1

File: c#/leetcode/15-3-sum.cs
public class Solution
{
  public IList<IList<int>> ThreeSum(int[] nums)
  {
    Array.Sort(nums);
    List<IList<int>> res = new List<IList<int>>();

    for (int i = 0; i < nums.Length; i++)
    {
      if (nums[i] > 0)
      {
        break;
      }

      if (i == 0 || nums[i] != nums[i - 1])
      {
        twoSums(nums, i, res);
      }
    }

    return res;
  }

  public void twoSums(int[] nums, int i, List<IList<int>> res)
  {
    int left = i + 1;
    int right = nums.Length - 1;

    while (left < right)
    {
      if (nums[left] + nums[right] + nums[i] == 0)
      {
        res.Add(new int[] { nums[left], nums[right], nums[i] });
        left++;
        right--;
        while (left < right && nums[left] == nums[left - 1])
        {
          left++;
        }
      }
      else if (nums[left] + nums[right] + nums[i] > 0)
      {
        right--;
      }
      else
      {
        left++;
      }
    }


  }
}